### DISCUSSION TITLE
Encryption key & Access tokens

### DISCUSSION THREAD

**Discussion by srivaths** [teslamate-org/teslamate#3220] [Q&A] [answered]
I just upgraded to 1.27.2.  I am now cautioned to specify an `ENCRYPTION_KEY`.  I have read the docs on where to specify the key (https://docs.teslamate.org/docs/installation/docker) and seen other threads about how to address this issue (such as https://github.com/adriankumpf/teslamate/discussions/2798).  I understand enough that 

* the key is used to access the token
* You have to generate the token (and refresh token)

What I _don't_ know is how and where to store these tokens.  Maybe I am having a stupid moment.  Can you please point me next steps once I have the tokens?  And is there any guidance on generating the key itself - or is it just like a strong passphrase?

Sri

**Comment by bodosom** [comment, marked as answer]
If you mean "**Choose a secure encryption key that will be used to encrypt your Tesla API tokens (insert as ENCRYPTION_KEY)**" it's in the yml file. If you're asking about making the API keys available to TeslaMate after you generate them you'll be prompted to provide them when you connect to port 4000 if the keys are missing, invalid or you cannot connect to the auth server. The two API keys are stored in the database and are encrypted using "ENCRYPTION_KEY".

**Reply by srivaths** [reply]
Thanks, @bodosom.  Your pointer helped. 

For anyone else running into the same issue here's what **I** had to do.

I read [docker.md](https://github.com/adriankumpf/teslamate/blob/master/website/docs/installation/docker.md), I realized that I needed to Update my `docker-compose.yml`.  I had to 
* Add an `ENCRYPTION_KEY` variable to the `environment` section and 
* Provide a `ports` section.  

Then I restarted Teslamate (using `docker-compose up -d`) 

When you access Teslamate via your browser you will be prompted to specify the API token and refresh token 
![image](https://github.com/adriankumpf/teslamate/assets/27290/ad05d338-57fc-48c0-b62b-dda4c634f1b2)

You can get this any number of ways.  I used the handy [tesla_auth tool](https://github.com/adriankumpf/tesla_auth).

Everything is good now.  Thanks

**Reply by solelo** [reply]
Where did you get the encryption key? I'm I have the variable in there already and it's still requiring me to put in a key, I'm just not sure where to get the key from

**Comment by sparvata** [comment]
when you browse to you dashboards, on the left panel there should be a gear icon. Click on that and once the Configuration page opens, Go to the API keys section and generate one based on your need. something like http://your-ip-address:3000/org/apikeys

**Reply by solelo** [reply]
Got it and it works. Thank you very much.

**Comment by BigKenD** [comment]
Still trying to find this... I am not using docker, but a straight install on Ubuntu.
Gear Icon takes me to Administration
There is no 'Configuration page'
If I go to the /org/apikeys
I get 
No API keys found
No API keys were found. If you reload the browser, this tab will be not available. If any API keys are created, this tab will appear again.

**Reply by BigKenD** [reply]
Found it in the end for people who DO NOT USE DOCKER....

The new installation instructions add a line to the /etc/systemd/system/teslamate.service file.

Environment="PORT=4000"
****Environment="ENCRYPTION_KEY=your_secure_encryption_key_here"****
Environment="DATABASE_USER=teslamate"

If you edit the file and inset "yoursecretkey" 

and run :-  

systemctl status teslamate

**_Warning: The unit file, source configuration file or drop-ins of teslamate.service changed on disk. 
Run 'systemctl daemon-reload' to reload units. ● teslamate.service - TeslaMate_**


you get the above.

Running :- 'systemctl daemon-reload'

You get :-
Aug 10 11:35:01 ubuntu0 teslamate[2278]: 2025-08-10 11:35:01.321 [info] POST **_https://auth.tesla.com/oauth2/v3/token_** -> 200 (262.799 ms)
Aug 10 11:35:01 ubuntu0 teslamate[2278]: 2025-08-10 11:35:01.397 [info] **_Scheduling token refresh in 6 h_**

Hope this helps people who have installed Teslamate without Docker.
